### PR TITLE
Further encapsulate allocation of slots 

### PR DESCRIPTION
**Pull request by zentol**
Based on #11.

Further encapsulates the reservation of slots by moving the reservation/free calls, along with usages of the requirements/mapping/calculators into a `SlotAllocator`(couldn't think of a better name).
Also moves all these things into a separate sub-package.

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeSchedulerNG.java [hunk 1]
```diff
@@ -764,72 +762,22 @@ private State createExecutionGraphWithAvailableResources() {
             }
         }
 
-        private DeclarativeScheduler.ParallelismAndResourceAssignments
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeSchedulerNG.java]: This should become a top-level class at some point.

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeSchedulerNG.java [hunk 2]
```diff
@@ -156,9 +151,7 @@
 
     @Nullable private JobStatusListener jobStatusListener;
 
-    private final RequirementsCalculator requirementsCalculator =
-            new SlotSharingRequirementsCalculator();
-    private final MappingCalculator mappingCalculator = new SlotSharingMappingCalculator();
+    private final SlotSharingSlotAllocator slotAllocator;
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeSchedulerNG.java:154]: I guess that we need to introduce a `SlotAllocator` interface at some point.

**Review by tillrohrmann — APPROVED** [review]
Thanks for creating this PR @zentol. LGTM. Merging.

**Comment by tillrohrmann** [comment]
Merged via 7005db486af86e455165e95b8f6c3415ec694e87

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeSchedulerNG.java (hunk 2 above)
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeSchedulerNG.java:154]: there is an interface, i just forgot to use it here

**Pull request closed without merging**